In [2]:
import pandas as pd
import os

In [3]:
non = pd.read_csv('non_cancerous_img.csv')
can = pd.read_csv('cancerous_images.csv')

In [4]:
non.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1218 entries, 0 to 1217
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Unnamed: 0               1218 non-null   int64  
 1   site_id                  1218 non-null   int64  
 2   patient_id               1218 non-null   int64  
 3   image_id                 1218 non-null   int64  
 4   laterality               1218 non-null   object 
 5   view                     1218 non-null   object 
 6   age                      1218 non-null   float64
 7   cancer                   1218 non-null   int64  
 8   biopsy                   1218 non-null   int64  
 9   invasive                 1218 non-null   int64  
 10  BIRADS                   1218 non-null   float64
 11  implant                  1218 non-null   int64  
 12  density                  1218 non-null   object 
 13  machine_id               1218 non-null   int64  
 14  difficult_negative_case 

In [5]:
can.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1158 entries, 0 to 1157
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Unnamed: 0               1158 non-null   int64  
 1   site_id                  1158 non-null   int64  
 2   patient_id               1158 non-null   int64  
 3   image_id                 1158 non-null   int64  
 4   laterality               1158 non-null   object 
 5   view                     1158 non-null   object 
 6   age                      1158 non-null   float64
 7   cancer                   1158 non-null   int64  
 8   biopsy                   1158 non-null   int64  
 9   invasive                 1158 non-null   int64  
 10  BIRADS                   664 non-null    float64
 11  implant                  1158 non-null   int64  
 12  density                  664 non-null    object 
 13  machine_id               1158 non-null   int64  
 14  difficult_negative_case 

In [5]:
# Concatenate the two DataFrames and reset the index
df = pd.concat([non, can], ignore_index=True).reset_index(drop=True)
df.head()

Unnamed: 0  site_id  patient_id    image_id laterality view   age  cancer  \
0         102        1        1014   629904228          L  MLO  76.0       0   
1         103        1        1014   669597068          L   CC  76.0       0   
2         104        1        1014   229558076          R  MLO  76.0       0   
3         105        1        1014  1173679750          R   CC  76.0       0   
4         166        1       10208   638273415          L  MLO  56.0       0   

   biopsy  invasive  BIRADS  implant density  machine_id  \
0       0         0     1.0        0       B          49   
1       0         0     1.0        0       B          49   
2       0         0     1.0        0       B          49   
3       0         0     1.0        0       B          49   
4       0         0     0.0        0       C          49   

   difficult_negative_case  
0                    False  
1                    False  
2                    False  
3                    False  
4                     True

In [6]:
df.drop('Unnamed: 0', axis = 1, inplace = True)
df.head()

site_id  patient_id    image_id laterality view   age  cancer  biopsy  \
0        1        1014   629904228          L  MLO  76.0       0       0   
1        1        1014   669597068          L   CC  76.0       0       0   
2        1        1014   229558076          R  MLO  76.0       0       0   
3        1        1014  1173679750          R   CC  76.0       0       0   
4        1       10208   638273415          L  MLO  56.0       0       0   

   invasive  BIRADS  implant density  machine_id  difficult_negative_case  
0         0     1.0        0       B          49                    False  
1         0     1.0        0       B          49                    False  
2         0     1.0        0       B          49                    False  
3         0     1.0        0       B          49                    False  
4         0     0.0        0       C          49                     True

In [7]:
df.head()

site_id  patient_id    image_id laterality view   age  cancer  biopsy  \
0        1        1014   629904228          L  MLO  76.0       0       0   
1        1        1014   669597068          L   CC  76.0       0       0   
2        1        1014   229558076          R  MLO  76.0       0       0   
3        1        1014  1173679750          R   CC  76.0       0       0   
4        1       10208   638273415          L  MLO  56.0       0       0   

   invasive  BIRADS  implant density  machine_id  difficult_negative_case  
0         0     1.0        0       B          49                    False  
1         0     1.0        0       B          49                    False  
2         0     1.0        0       B          49                    False  
3         0     1.0        0       B          49                    False  
4         0     0.0        0       C          49                     True

In [8]:
# Filter the DataFrame to get rows where 'view' is 'AT'
filtered_df = df[df['view'] == 'AT']
filtered_df

site_id  patient_id    image_id laterality view   age  cancer  biopsy  \
1500        1       25323   106495931          R   AT  74.0       1       1   
1504        1       25323  1743461841          R   AT  74.0       1       1   

      invasive  BIRADS  implant density  machine_id  difficult_negative_case  
1500         1     0.0        0       B          49                    False  
1504         1     0.0        0       B          49                    False

In [9]:
# Iimage with view == 'AT'
image_paths = []
for i, row in filtered_df.iterrows():
    # Get the image_id from the column
    image_id = row['image_id']
    
    # Construct the file path (assuming images are in DICOM format, e.g., '.dcm')
    image_path = os.path.join('IMG', f"{image_id}.dcm")  # Modify the extension if needed
    image_paths.append(image_path)
image_paths

['IMG\\106495931.dcm', 'IMG\\1743461841.dcm']

### **CODE TO DLT IMG BELOW**

In [18]:
# Remove  images for which view == 'AT'
for image_path in image_paths:
    # Check if the file exists
    if os.path.exists(image_path):
        os.remove(image_path)  # Delete the image file
        print(f"Deleted: {image_path}")
    else: 
        print(f"File not found: {image_path}")

Deleted: IMG\106495931.dcm
Deleted: IMG\1743461841.dcm


In [10]:
# Delete those rows from the DataFrame
df = df[df['view'] != 'AT']

In [11]:
df.isna().sum()

site_id                      0
patient_id                   0
image_id                     0
laterality                   0
view                         0
age                          0
cancer                       0
biopsy                       0
invasive                     0
BIRADS                     494
implant                      0
density                    494
machine_id                   0
difficult_negative_case      0
dtype: int64

there are missing values in BIRADS and density. 
494 / 2374
494 missing values are for cancerous dataset.
cancer dataset has only 0 as value for the column BIRADS while non_cancerous has other categories 0,1,2
values in density also missing for the same records that has missing values in BIRADS.
These are critical values and it seems that it is not advisable to impute them by traditional methods.
that is so much. so,imputing might not be good. so a copy without those columns

In [15]:
df_copy = df.copy(deep = True)
df_copy.drop(columns = ['BIRADS', 'density'], inplace = True)

In [16]:
df_copy.head()

site_id  patient_id    image_id laterality view   age  cancer  biopsy  \
0        1        1014   629904228          L  MLO  76.0       0       0   
1        1        1014   669597068          L   CC  76.0       0       0   
2        1        1014   229558076          R  MLO  76.0       0       0   
3        1        1014  1173679750          R   CC  76.0       0       0   
4        1       10208   638273415          L  MLO  56.0       0       0   

   invasive  implant  machine_id  difficult_negative_case  
0         0        0          49                    False  
1         0        0          49                    False  
2         0        0          49                    False  
3         0        0          49                    False  
4         0        0          49                     True

In [28]:
# adding a new column image Path
df_copy['image_path'] = df_copy['image_id'].apply(lambda x: os.path.join('IMG', f'{x}.dcm'))
df_copy.head()

site_id  patient_id    image_id laterality view   age  cancer  biopsy  \
0        1        1014   629904228          L  MLO  76.0       0       0   
1        1        1014   669597068          L   CC  76.0       0       0   
2        1        1014   229558076          R  MLO  76.0       0       0   
3        1        1014  1173679750          R   CC  76.0       0       0   
4        1       10208   638273415          L  MLO  56.0       0       0   

   invasive  implant  machine_id  difficult_negative_case          image_path  
0         0        0          49                    False   IMG\629904228.dcm  
1         0        0          49                    False   IMG\669597068.dcm  
2         0        0          49                    False   IMG\229558076.dcm  
3         0        0          49                    False  IMG\1173679750.dcm  
4         0        0          49                     True   IMG\638273415.dcm

In [29]:
# save the fie
df_copy.to_csv('RSNA_breast_cancer_data.csv', index=False)